README

 Copyright (c) 2007 Graphcomp. All rights reserved.
 This program is free software; you can redistribute it and/or
 modify it under the same terms as Perl itself. 

This module provides methods that abstract and simplify loading/managing
OpenGL shaders.

It requires the OpenGL module - version 0.56 or newer.

You can find the latest releases and information for installing/using
these modules at the Perl OpenGL (POGL) Developer's Site:
http://graphcomp.com/opengl



Changes:

1.0	First release - Bob "grafman" Free <grafman@graphcomp.com>


Changes
Change List for 1.00 - Bob "grafman" Free <bfree@graphcomp.com>
____________________________________________________________

New Changes

New Copyright

New INSTALL

New Makefile.PL

New MANIFEST

New README

New Shader.pm

New Shader.pod

New Shader/Common.pm

New Shader/Objects.pm

New Shader/ARB.pm

New Shader/GLSL.pm

New Shader/CG.pm

New t/OpenGL-Shader.t

New fragment.arb

New vertex.arb

New fragment.glsl

New vertex.glsl

New fragment.cg

New vertex.cg




Shader/CG.pm
############################################################
#
# OpenGL::Shader::CG - Copyright 2007 Graphcomp - ALL RIGHTS RESERVED
# Author: Bob "grafman" Free - grafman@graphcomp.com
#
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.
#
############################################################

package OpenGL::Shader::CG;

require Exporter;

use Carp;

use vars qw($VERSION $DESCRIPTION @ISA);
$VERSION = '1.00';

$DESCRIPTION = qq
{nVidia's Cg Shader Language};

use OpenGL::Shader::Objects;
@ISA = qw(Exporter OpenGL::Shader::Objects);

use OpenGL(':all');



=head1 NAME

  OpenGL::Shader::CG - copyright 2007 Graphcomp - ALL RIGHTS RESERVED
  Author: Bob "grafman" Free - grafman@graphcomp.com

  This program is free software; you can redistribute it and/or
  modify it under the same terms as Perl itself.


=head1 DESCRIPTION

  This is a plug-in module for use with the OpenGL::Shader.
  While it may be called directly, it will more often be called
  by the OpenGL::Shader abstraction module.

  This is a subclass of the OpenGL::Shader::Objects module.


=head1 SYNOPSIS

  ##########
  # Instantiate a shader

  use OpenGL::Shader;
  my $shdr = new OpenGL::Shader('CG');


  ##########
  # Methods defined in OpenGL::Shader::Common:

  # Get shader type.
  my $ver = $shdr->GetType();

  # Load shader files.
  my $stat = $shdr->LoadFiles($fragment_file,$vertex_file);

  # Get shader GL constants.
  my $fragment_const = $shdr->GetFragmentConstant();
  my $vertex_const = $shdr->GetVertexConstant();

  # Get objects.
  my $fragment_shader = $shdr->GetFragmentShader();
  my $vertex_shader = $shdr->GetVertexShader();
  my $program = $shdr->GetProgram();


  ##########
  # Methods defined in OpenGL::Shader::Objects:

  # Load shader text.
  $shdr->Load($fragment,$vertex);

  # Enable shader.
  $shdr->Enable();

  # Set Vertex Attribute
  my $attr_id = $self->MapAttr($attr_name);
  glVertexAttrib4fARB($attr_id,$x,$y,$z,$w);

  # Get Global Variable ID (uniform/env)
  my $var_id = $self->Map($var_name);

  # Set float4 vector variable
  $stat = $self->SetVector($var_name,$x,$y,$z,$w);

  # Set float4x4 matrix via OGA
  $stat = $self->SetMatrix($var_name,$oga);

  # Disable shader.
  $shdr->Disable();

  # Destructor.
  $shdr->DESTROY();


  ##########
  # Methods defined in this module:

  # Get shader version.
  my $ver = $shdr->GetVersion();


=cut


# Shader constructor
sub new
{
  my $this = shift;
  my $class = ref($this) || $this;

  my $self = new OpenGL::Shader::Objects(@_);
  return undef if (!$self);
  bless($self,$class);

  # Check for additional required OpenGL extensions
  return undef if (OpenGL::glpCheckExtension('GL_EXT_Cg_shader'));

  $self->{type} = 'CG';

  $self->{fragment_const} = GL_CG_FRAGMENT_SHADER_EXT;
  $self->{vertex_const} = GL_CG_VERTEX_SHADER_EXT;

  return $self;
}


# GetVersion
sub GetVersion
{
  my($self,$shader_ver) = @_;

  if (!$self->{version})
  {
    # Get GL_SHADING_LANGUAGE_VERSION_ARB
    $shader_ver = glGetString(0x8B8C) if (!$shader_ver);

    return '1.00' if ($shader_ver !~ m|Cg ([\d\.]+)|i);
    $self->{version} = $1;
  }

  return $self->{version};
}


1;
__END__



Shader/ARB.pm
############################################################
#
# OpenGL::Shader::ARB - Copyright 2007 Graphcomp - ALL RIGHTS RESERVED
# Author: Bob "grafman" Free - grafman@graphcomp.com
#
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.
#
############################################################

package OpenGL::Shader::ARB;

require Exporter;

use Carp;

use vars qw($VERSION $DESCRIPTION @ISA);
$VERSION = '1.00';

$DESCRIPTION = qq
{Supports ARBfp1.0 and ARBvp1.0};

use OpenGL::Shader::Common;
@ISA = qw(Exporter OpenGL::Shader::Common);

use OpenGL(':all');



=head1 NAME

  OpenGL::Shader::ARB - copyright 2007 Graphcomp - ALL RIGHTS RESERVED
  Author: Bob "grafman" Free - grafman@graphcomp.com

  This program is free software; you can redistribute it and/or
  modify it under the same terms as Perl itself.


=head1 DESCRIPTION

  This is a plug-in module for use with the OpenGL::Shader.
  While it may be called directly, it will more often be called
  by the OpenGL::Shader abstraction module.

  This is a subclass of the OpenGL::Shader::Common module.


=head1 SYNOPSIS

  ##########
  # Instantiate a shader

  use OpenGL::Shader;
  my $shdr = new OpenGL::Shader('ARB');


  ##########
  # Methods defined in OpenGL::Shader::Common:

  # Get shader type.
  my $ver = $shdr->GetType();

  # Load shader files.
  my $stat = $shdr->LoadFiles($fragment_file,$vertex_file);

  # Get shader GL constants.
  my $fragment_const = $shdr->GetFragmentConstant();
  my $vertex_const = $shdr->GetVertexConstant();

  # Get objects.
  my $fragment_shader = $shdr->GetFragmentShader();
  my $vertex_shader = $shdr->GetVertexShader();
  my $program = $shdr->GetProgram();


  ##########
  # Methods defined in this module:

  # Get shader version.
  my $ver = $shdr->GetVersion();

  # Load shader text.
  $shdr->Load($fragment,$vertex);

  # Enable shader.
  $shdr->Enable();

  # Set Vertex Attribute
  my $attr_id = $self->MapAttr($attr_name);
  glVertexAttrib4fARB($attr_id,$x,$y,$z,$w);

  # Get Global Variable ID (uniform/env)
  my $var_id = $self->Map($var_name);

  # Set float4 vector variable
  $stat = $self->SetVector($var_name,$x,$y,$z,$w);

  # Set float4x4 matrix via OGA
  $stat = $self->SetMatrix($var_name,$oga);

  # Disable shader.
  $shdr->Disable();

  # Destructor.
  $shdr->DESTROY();


=cut


# Shader constructor
sub new
{
  my $this = shift;
  my $class = ref($this) || $this;

  my $self = new OpenGL::Shader::Common(@_);
  return undef if (!$self);
  bless($self,$class);

  # Check for required OpenGL extensions
  return undef if (OpenGL::glpCheckExtension('GL_ARB_fragment_program'));
  return undef if (OpenGL::glpCheckExtension('GL_ARB_vertex_program'));

  $self->{type} = 'ARB';
  $self->{version} = '1.00';

  $self->{fragment_const} = GL_FRAGMENT_PROGRAM_ARB;
  $self->{vertex_const} = GL_VERTEX_PROGRAM_ARB;

  ($self->{fragment_id},$self->{vertex_id}) = glGenProgramsARB_p(2);
  return undef if (!$self->{fragment_id} || !$self->{vertex_id});

  return $self;
}


# Shader destructor
# Must be disabled first
sub DESTROY
{
  my($self) = @_;
  glDeleteProgramsARB_p($self->{fragment_id}) if ($self->{fragment_id});
  glDeleteProgramsARB_p($self->{vertex_id}) if ($self->{vertex_id});
}


# GetVersion
sub GetVersion
{
  my($self,$shader_ver) = @_;

  # ARBfp1.0 and ARBvp1.0
  return $self->{version};
}


# Load shader strings
sub Load
{
  my($self,$fragment,$vertex) = @_;

  glBindProgramARB($self->{fragment_const}, $self->{fragment_id});
  glProgramStringARB_p($self->{fragment_const}, $fragment);
  $self->{fragment_code} = $fragment;
  $self->{frag_vars} = {};

  glBindProgramARB($self->{vertex_const}, $self->{vertex_id});
  glProgramStringARB_p($self->{vertex_const}, $vertex);
  $self->{vertex_code} = $vertex;
  $self->{vert_vars} = {};

  return '';
}


# Enable shader
sub Enable
{
  my($self) = @_;

  glEnable($self->{fragment_const});
  glEnable($self->{vertex_const});
}


# Disable shader
sub Disable
{
  my($self) = @_;

  glDisable($self->{fragment_const});
  glDisable($self->{vertex_const});
}


# Return shader vertex attribute ID
sub MapAttr
{
  my($self,$attr) = @_;
  return undef if ($self->load_attrs());
  return $self->{attrs}->{$attr};
}


# Return shader local variable ID
sub Map
{
  my($self,$var,$prog) = @_;

  $self->load_vars($self->{frag_vars},$self->{fragment_code});
  $self->load_vars($self->{vert_vars},$self->{vertex_code});

  if ($prog == 1)
  {
    return undef if (!$self->{frag_vars});
    return $self->{frag_vars}->{$var};
  }

  if ($prog == 2)
  {
    return undef if (!$self->{vert_vars});
    return $self->{vert_vars}->{$var};
  }

  if ($self->{vert_vars})
  {
    my $id = $self->{vert_vars}->{$var};
    return $id if (defined($id));
  }

  return undef if (!$self->{frag_vars});
  return $self->{frag_vars}->{$var};
}


# Set shader vector
sub SetVector
{
  my($self,$var,@values) = @_;

  my $count = scalar(@values);
  return 'Invalid number of values' if ($count != 4);

  my $id = $self->Map($var,2);
  if (defined($id))
  {
    glProgramLocalParameter4fARB($self->{vertex_const},$id,@values);
    return '';
  }

  $id = $self->Map($var,1);
  return 'Unable to map $var' if (!defined($id));

  glProgramLocalParameter4fARB($self->{fragment_const},$id,@values);
  return '';
}


# Set shader 4x4 matrix
sub SetMatrix
{
  my($self,$var,$oga) = @_;
  return 'No oga supplied' if (!$oga);

  my $id = $self->Map($var,2);
  if (defined($id))
  {
    $self->set_matrix($self->{vertex_const},$id,$oga);
    return '';
  }

  $id = $self->Map($var,1);
  return 'Unable to map $var' if (!defined($id));

  $self->set_matrix($self->{fragment_const},$id,$oga);
  return '';
}

sub set_matrix
{
  my($self,$const,$id,$oga) = @_;
  for (my $i=0; $i<4; $i++)
  {
    glProgramLocalParameter4fvARB_c($const,$id+$i,$oga->offset($i*4));
  }
}


# Parse attribute names
sub load_attrs
{
  my($self) = @_;
  return '' if ($self->{attrs});
  return 'No vertex program' if (!$self->{vertex_code});

  my @lines = split('[\r\n]+',$self->{vertex_code});
  foreach my $line (@lines)
  {
    next if ($line !~ m|ATTRIB\s*(\w+)\s*\=\s*vertex\.([^;]+)|);
    my($var,$map) = ($1,$2);

    if ($map =~ m|attrib\[(\d+)\]|)
    {
      $self->{attrs}->{$var} = $1;
    }
    elsif ($map eq 'position')
    {
      $self->{attrs}->{$var} = 0;
    }
    elsif ($map =~ m|weight|)
    {
      $self->{attrs}->{$var} = 1;
    }
    elsif ($map eq 'normal')
    {
      $self->{attrs}->{$var} = 2;
    }
    elsif ($map eq 'color.secondary')
    {
      $self->{attrs}->{$var} = 4;
    }
    elsif ($map =~ m|color|)
    {
      $self->{attrs}->{$var} = 3;
    }
    elsif ($map =~ m|fogcoord|)
    {
      $self->{attrs}->{$var} = 5;
    }
    elsif ($map eq 'texcoord')
    {
      $self->{attrs}->{$var} = 8;
    }
    elsif ($map =~ m|texcoord\[(\d+)\]|)
    {
      $self->{attrs}->{$var} = $1;
    }
  }
}

# Parse variable names
sub load_vars
{
  my($self,$href,$code) = @_;
  return '' if (scalar(%$href));
  return 'No program' if (!$code);

  my @lines = split('[\r\n]+',$code);
  foreach my $line (@lines)
  {
    next if ($line !~ m|PARAM\s*([^\[\s]+)[\[\d\]\s]*\=[\{\s]*program\.local\[([0-9]+)|);
    my($var,$index) = ($1,$2);
    $href->{$var} = $index;
  }
}


1;
__END__



Shader/GLSL.pm
############################################################
#
# OpenGL::Shader::GLSL - Copyright 2007 Graphcomp - ALL RIGHTS RESERVED
# Author: Bob "grafman" Free - grafman@graphcomp.com
#
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.
#
############################################################

package OpenGL::Shader::GLSL;

require Exporter;

use Carp;

use vars qw($VERSION $DESCRIPTION @ISA);
$VERSION = '1.00';

$DESCRIPTION = qq
{OpenGL Shader Language};

use OpenGL::Shader::Objects;
@ISA = qw(Exporter OpenGL::Shader::Objects);

use OpenGL(':all');



=head1 NAME

  OpenGL::Shader::GLSL - copyright 2007 Graphcomp - ALL RIGHTS RESERVED
  Author: Bob "grafman" Free - grafman@graphcomp.com

  This program is free software; you can redistribute it and/or
  modify it under the same terms as Perl itself.


=head1 DESCRIPTION

  This is a plug-in module for use with the OpenGL::Shader.
  While it may be called directly, it will more often be called
  by the OpenGL::Shader abstraction module.

  This is a subclass of the OpenGL::Shader::Objects module.


=head1 SYNOPSIS

  ##########
  # Instantiate a shader

  use OpenGL::Shader;
  my $shdr = new OpenGL::Shader('GLSL');


  ##########
  # Methods defined in OpenGL::Shader::Common:

  # Get shader type.
  my $ver = $shdr->GetType();

  # Load shader files.
  my $stat = $shdr->LoadFiles($fragment_file,$vertex_file);

  # Get shader GL constants.
  my $fragment_const = $shdr->GetFragmentConstant();
  my $vertex_const = $shdr->GetVertexConstant();

  # Get objects.
  my $fragment_shader = $shdr->GetFragmentShader();
  my $vertex_shader = $shdr->GetVertexShader();
  my $program = $shdr->GetProgram();


  ##########
  # Methods defined in OpenGL::Shader::Objects:

  # Load shader text.
  $shdr->Load($fragment,$vertex);

  # Enable shader.
  $shdr->Enable();

  # Set Vertex Attribute
  my $attr_id = $self->MapAttr($attr_name);
  glVertexAttrib4fARB($attr_id,$x,$y,$z,$w);

  # Get Global Variable ID (uniform/env)
  my $var_id = $self->Map($var_name);

  # Set float4 vector variable
  $stat = $self->SetVector($var_name,$x,$y,$z,$w);

  # Set float4x4 matrix via OGA
  $stat = $self->SetMatrix($var_name,$oga);

  # Disable shader.
  $shdr->Disable();

  # Destructor.
  $shdr->DESTROY();


  ##########
  # Methods defined in this module:

  # Get shader version.
  my $ver = $shdr->GetVersion();


=cut


# Shader constructor
sub new
{
  my $this = shift;
  my $class = ref($this) || $this;

  my $self = new OpenGL::Shader::Objects(@_);
  return undef if (!$self);
  bless($self,$class);

  $self->{type} = 'GLSL';

  $self->{fragment_const} = GL_FRAGMENT_SHADER;
  $self->{vertex_const} = GL_VERTEX_SHADER;

  return $self;
}


# GetVersion
sub GetVersion
{
  my($self,$shader_ver) = @_;

  if (!$self->{version})
  {
    # Get GL_SHADING_LANGUAGE_VERSION_ARB
    $shader_ver = glGetString(0x8B8C) if (!$shader_ver);

    return '1.00' if ($shader_ver !~ m|([\d\.]+)|);
    $self->{version} = $1;
  }

  return $self->{version};
}


1;
__END__



Shader/Common.pm
############################################################
#
# OpenGL::Shader::Common - Copyright 2007 Graphcomp - ALL RIGHTS RESERVED
# Author: Bob "grafman" Free - grafman@graphcomp.com
#
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.
#
############################################################

package OpenGL::Shader::Common;

require Exporter;

use Carp;

use vars qw($VERSION @ISA);
$VERSION = '1.00';

@ISA = qw(Exporter);


=head1 NAME

  OpenGL::Shader::Common - copyright 2007 Graphcomp - ALL RIGHTS RESERVED
  Author: Bob "grafman" Free - grafman@graphcomp.com

  This program is free software; you can redistribute it and/or
  modify it under the same terms as Perl itself.


=head1 DESCRIPTION

  This module provides a base class for OpenGL shader types.
  Requires the OpenGL module.


=head1 SYNOPSIS

  ##########
  # Instantiate a shader

  use OpenGL::Shader;
  my $shdr = new OpenGL::Shader('GLSL');


  ##########
  # Methods defined in this module:

  # Get shader type.
  my $ver = $shdr->GetType();

  # Load shader files.
  my $stat = $shdr->LoadFiles($fragment_file,$vertex_file);

  # Get shader GL constants.
  my $fragment_const = $shdr->GetFragmentConstant();
  my $vertex_const = $shdr->GetVertexConstant();

  # Get objects.
  my $fragment_shader = $shdr->GetFragmentShader();
  my $vertex_shader = $shdr->GetVertexShader();
  my $program = $shdr->GetProgram();


  ##########
  # Methods defined in subclasses:

  # Get shader version.
  my $ver = $shdr->GetVersion();

  # Load shader text.
  $shdr->Load($fragment,$vertex);

  # Enable shader.
  $shdr->Enable();

  # Set Vertex Attribute
  my $attr_id = $self->MapAttr($attr_name);
  glVertexAttrib4fARB($attr_id,$x,$y,$z,$w);

  # Get Global Variable ID (uniform/env)
  my $var_id = $self->Map($var_name);

  # Set float4 vector variable
  $stat = $self->SetVector($var_name,$x,$y,$z,$w);

  # Set float4x4 matrix via OGA
  $stat = $self->SetMatrix($var_name,$oga);

  # Disable shader.
  $shdr->Disable();

  # Destructor.
  $shdr->DESTROY();

=cut


# Base constructor
sub new
{
  my $this = shift;
  my $class = ref($this) || $this;
  my($type) = @_;
  my $self = {type => uc($type)};
  bless($self,$class);

  return $self;
}


# Get shader type
sub GetType
{
  my($self) = @_;
  return $self->{type};
}


# Get fragment shader constant
sub GetFragmentConstant
{
  my($self) = @_;
  return $self->{fragment_const};
}


# Get vertex shader constant
sub GetVertexConstant
{
  my($self) = @_;
  return $self->{vertex_const};
}


# Get fragment shader ID
sub GetFragmentShader
{
  my($self) = @_;
  return $self->{fragment_id};
}


# Get vertex shader ID
sub GetVertexShader
{
  my($self) = @_;
  return $self->{vertex_id};
}


# Get shader program object
sub GetProgram
{
  my($self) = @_;
  return $self->{program};
}


# Load shader files
sub LoadFiles
{
  my($self,$fragment_file,$vertex_file) = @_;

  my $fragment = '';
  if ($fragment_file)
  {
    return "Does not exist: $fragment_file" if (!-e $fragment_file);
    $fragment = $self->read_file($fragment_file);
    return "Empty fragment file" if (!$fragment);
  }

  my $vertex = '';
  if ($fragment_file)
  {
    return "Does not exist: $vertex_file" if (!-e $vertex_file);
    $vertex = $self->read_file($vertex_file);
    return "Empty vertex file" if (!$vertex);
  }

  return $self->Load($fragment,$vertex);
}


# Read file
sub read_file
{
  my($self,$file) = @_;

  return undef if (!open(FILE,$file));
  my @data = <FILE>;
  close(FILE);
  return join('',@data);
}


1;
__END__



Shader/Objects.pm
############################################################
#
# OpenGL::Shader::Objects - Copyright 2007 Graphcomp - ALL RIGHTS RESERVED
# Author: Bob "grafman" Free - grafman@graphcomp.com
#
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.
#
############################################################

package OpenGL::Shader::Objects;

require Exporter;

use Carp;

use vars qw($VERSION $DESCRIPTION @ISA);
$VERSION = '1.00';

use OpenGL::Shader::Common;
@ISA = qw(Exporter OpenGL::Shader::Common);

use OpenGL(':all');



=head1 NAME

  OpenGL::Shader::Objects - copyright 2007 Graphcomp - ALL RIGHTS RESERVED
  Author: Bob "grafman" Free - grafman@graphcomp.com

  This program is free software; you can redistribute it and/or
  modify it under the same terms as Perl itself.


=head1 DESCRIPTION

  This is a base class for plug-in modules based on GL_ARB_shader_objects.
  This is a subclass of the OpenGL::Shader::Common module.


=head1 SYNOPSIS

  ##########
  # Instantiate a shader

  use OpenGL::Shader;
  my $shdr = new OpenGL::Shader('CG');


  ##########
  # Methods defined in OpenGL::Shader::Common:

  # Get shader type.
  my $ver = $shdr->GetType();

  # Load shader files.
  my $stat = $shdr->LoadFiles($fragment_file,$vertex_file);

  # Get shader GL constants.
  my $fragment_const = $shdr->GetFragmentConstant();
  my $vertex_const = $shdr->GetVertexConstant();

  # Get objects.
  my $fragment_shader = $shdr->GetFragmentShader();
  my $vertex_shader = $shdr->GetVertexShader();
  my $program = $shdr->GetProgram();


  ##########
  # Methods defined in this module:

  # Load shader text.
  $shdr->Load($fragment,$vertex);

  # Enable shader.
  $shdr->Enable();

  # Set Vertex Attribute
  my $attr_id = $self->MapAttr($attr_name);
  glVertexAttrib4fARB($attr_id,$x,$y,$z,$w);

  # Get Global Variable ID (uniform/env)
  my $var_id = $self->Map($var_name);

  # Set float4 vector variable
  $stat = $self->SetVector($var_name,$x,$y,$z,$w);

  # Set float4x4 matrix via OGA
  $stat = $self->SetMatrix($var_name,$oga);

  # Disable shader.
  $shdr->Disable();

  # Destructor.
  $shdr->DESTROY();


  ##########
  # Methods defined in sub-class:

  # Get shader version.
  my $ver = $shdr->GetVersion();


=cut


# Shader constructor
sub new
{
  my $this = shift;
  my $class = ref($this) || $this;

  my $self = new OpenGL::Shader::Common(@_);
  return undef if (!$self);
  bless($self,$class);

  # Check for required OpenGL extensions
  return undef if (OpenGL::glpCheckExtension('GL_ARB_shader_objects'));
  return undef if (OpenGL::glpCheckExtension('GL_ARB_fragment_shader'));
  return undef if (OpenGL::glpCheckExtension('GL_ARB_vertex_shader'));

  $self->{type} = '';
  $self->{version} = '';

  $self->{fragment_const} = '';
  $self->{vertex_const} = '';

  return $self;
}


# Shader destructor
# Must be disabled first
sub DESTROY
{
  my($self) = @_;

  if ($self->{program})
  {
    glDetachObjectARB($self->{program},$self->{fragment_id}) if ($self->{fragment_id});
    glDetachObjectARB($self->{program},$self->{vertex_id}) if ($self->{vertex_id});
    glDeleteProgramsARB_p($self->{program});
  }

  glDeleteProgramsARB_p($self->{fragment_id}) if ($self->{fragment_id});
  glDeleteProgramsARB_p($self->{vertex_id}) if ($self->{vertex_id});
}


# Load shader strings
sub Load
{
  my($self,$fragment,$vertex) = @_;

  # Load fragment code
  if ($fragment)
  {
    $self->{fragment_id} = glCreateShaderObjectARB($self->{fragment_const});
    return undef if (!$self->{fragment_id});

    glShaderSourceARB_p($self->{fragment_id}, $fragment);
    #my $frag = glGetShaderSourceARB_p($self->{fragment_id});
    #print STDERR "Loaded fragment:\n$frag\n";

    glCompileShaderARB($self->{fragment_id});
    my $stat = glGetInfoLogARB_p($self->{fragment_id});
    return "Fragment shader: $stat" if ($stat);
  }

  # Load vertex code
  if ($vertex)
  {
    $self->{vertex_id} = glCreateShaderObjectARB($self->{vertex_const});
    return undef if (!$self->{vertex_id});

    glShaderSourceARB_p($self->{vertex_id}, $vertex);
    #my $vert = glGetShaderSourceARB_p($self->{vertex_id});
    #print STDERR "Loaded vertex:\n$vert\n";

    glCompileShaderARB($self->{vertex_id});
    $stat = glGetInfoLogARB_p($self->{vertex_id});
    return "Vertex shader: $stat" if ($stat);
  }


  # Link shaders
  my $sp = glCreateProgramObjectARB();
  glAttachObjectARB($sp, $self->{fragment_id}) if ($fragment);
  glAttachObjectARB($sp, $self->{vertex_id}) if ($vertex);
  glLinkProgramARB($sp);
  my $linked = glGetObjectParameterivARB_p($sp, GL_OBJECT_LINK_STATUS_ARB);
  if (!$linked)
  {
    $stat = glGetInfoLogARB_p($sp);
    #print STDERR "Load shader: $stat\n";
    return "Link shader: $stat" if ($stat);
    return 'Unable to link shader';
  }

  $self->{program} = $sp;

  return '';
}


# Enable shader
sub Enable
{
  my($self) = @_;
  glUseProgramObjectARB($self->{program}) if ($self->{program});
}


# Disable shader
sub Disable
{
  my($self) = @_;
  glUseProgramObjectARB(0) if ($self->{program});
}


# Return shader vertex attribute ID
sub MapAttr
{
  my($self,$attr) = @_;
  return undef if (!$self->{program});
  my $id = glGetAttribLocationARB_p($self->{program},$attr);
  return undef if ($id < 0);
  return $id;
}


# Return shader uniform variable ID
sub Map
{
  my($self,$var) = @_;
  return undef if (!$self->{program});
  my $id = glGetUniformLocationARB_p($self->{program},$var);
  return undef if ($id < 0);
  return $id;
}


# Set shader vector
sub SetVector
{
  my($self,$var,@values) = @_;

  my $id = $self->Map($var);
  return 'Unable to map $var' if (!defined($id));

  my $count = scalar(@values);
  return 'Invalid number of values' if ($count != 4);

  glUniform4fARB($id,@values);
  return '';
}


# Set shader 4x4 matrix
sub SetMatrix
{
  my($self,$var,$oga) = @_;

  my $id = $self->Map($var);
  return 'Unable to map $var' if (!defined($id));

  glUniformMatrix4fvARB_c($id,1,0,$oga->ptr());
  return '';
}


1;
__END__



Shader.pm
############################################################
#
# OpenGL::Shader - Copyright 2007 Graphcomp - ALL RIGHTS RESERVED
# Author: Bob "grafman" Free - grafman@graphcomp.com
#
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.
#
############################################################

package OpenGL::Shader;

require Exporter;

use Carp;

use vars qw($VERSION $SHADER_TYPES @ISA);
$VERSION = '1.00';

@ISA = qw(Exporter);

use OpenGL(':all');


=head1 NAME

  OpenGL::Shader - copyright 2007 Graphcomp - ALL RIGHTS RESERVED
  Author: Bob "grafman" Free - grafman@graphcomp.com

  This program is free software; you can redistribute it and/or
  modify it under the same terms as Perl itself.


=head1 DESCRIPTION

  This module provides an extensible abstraction for managing OpenGL shaders.

  Requires OpenGL v0.56 or newer.


=head1 SYNOPSIS


  ##########
  # Load and manage a shader
  use OpenGL(':all');
  use OpenGL::Shader;

  # Note: these APIs/methods requires a current GL context/window
  glutInit();
  glutInitDisplayMode(GLUT_RGBA);
  glutInitWindowSize($w,$h);
  my $Window_ID = glutCreateWindow( "OpenGL::Shader test" );

  # Get a hashref of shader types
  my $shaders = $OpenGL::Shader::GetTypes();

  # Test for shader support
  my $ok = OpenGL::Shader::HasType('Cg');


  # Instantiate a shader - list acceptable shaders by priority
  my $shdr = new OpenGL::Shader('GLSL','ARB');

  # Get shader type and version
  my $type = $shdr->GetType();
  my $ver = $shdr->GetVersion();

  # Load shader by strings
  my $stat = $shdr->Load($fragment,$vertex);

  # Load shader by file
  my $stat = $shdr->LoadFiles($fragment_file,$vertex_file);

  # Enable
  $shdr->Enable();


  # Get vertex attribute ID
  # returns undef if not supported or not found
  my $attr_id = $self->MapAttr($attr_name);
  glVertexAttrib4fARB($attr_id,$x,$y,$z,$w);

  # Get Global Variable ID (uniform/local)
  my $var_id = $self->Map($var_name);

  # Set float4 vector variable
  $stat = $self->SetVector($var_name,$x,$y,$z,$w);

  # Set float4x4 matrix via OGA
  $stat = $self->SetMatrix($var_name,$oga);


  # Do GL rendering


  # Disable
  $shdr->Disable();

  # Done
  glutDestroyWindow($Window_ID);

=cut



# Return hashref of supported shader types.
# Must be able to opens a hidden GL context.
# Use OpenGL/Shader/Types.lst if exists.
sub GetTypes
{
  return $SHADER_TYPES if (ref($SHADER_TYPES));

  my $dir = __FILE__;
  return undef if ($dir !~ s|\.pm$||);

  @types = ();
  # Use type list if exists
  my $list = "$dir/Types.lst";
  if (open(LIST,$list))
  {
    foreach my $type (<LIST>)
    {
      $type =~ s|[\r\n]+||g;
      next if (!-e "$dir/$type.pm");
      push(@types,$type);
    }
    close(LIST);
  }
  # Otherwise grab OpenGL/Shader modules
  elsif (opendir(DIR,$dir))
  {
    foreach my $type (readdir(DIR))
    {
      next if ($type !~ s|\.pm$||);
      push(@types,$type);
    }
    closedir(DIR);
  }
  return undef if (!scalar(@types));

  # Get GL_SHADING_LANGUAGE_VERSION_ARB
  my $shader_ver = glGetString(0x8B8C);

  $SHADER_TYPES = {};
  # Get module versions
  foreach my $type (@types)
  {
    next if ($type eq 'Common');
    next if ($type eq 'Objects');
    my $version = GetTypeVersion($type,$shader_ver);
    next if (!$version);
    $SHADER_TYPES->{$type} = $version;
  }

  return $SHADER_TYPES;
}


# Get shader's version
# Must be able to opens a hidden GL context.
sub GetTypeVersion
{
  my($type,$ver) = @_;
  return undef if (!$type);

  my $obj;
  my $module = GetTypeModule($type);

  my $exec = qq
  {
    use $module;
    \$obj = new $module\();
  };
  eval($exec);
  return undef if (!$obj || $@);

  return $obj->GetVersion($ver);
}


# Check for engine availability; returns installed version
sub HasType
{
  my($type,$min_ver,$max_var) = @_;

  my $types = GetTypes();
  return undef if (!$types);

  my $ver = $types->{$type};
  return undef if (!$ver);
  return undef if ($min_ver && $ver lt $min_ver);
  return undef if ($max_ver && $ver gt $max_ver);

  return $ver;
}


# Constructor wrapper for shader type
sub new
{
  my $this = shift;
  my $class = ref($this) || $this;
  my $self = {};
  bless($self,$class);

  my @types = @_ ? @_ : ('GLSL','CG','ARB');
  foreach my $type (@types)
  {
    my $obj;
    my $module = GetTypeModule($type);
    my $exec = qq
    {
      use $module;
      \$obj = new $module;
    };
    eval($exec);

    return $obj if ($obj && !$@);
  }
  return undef;
}


# Get shader module name
sub GetTypeModule
{
  my($type) = @_;
  return __PACKAGE__.'::'.uc($type);
}




1;
__END__

